File: core/src/com/questmark/entity/systems/collision/EntityCollisionSystem.java
package com.questmark.entity.systems.collision;

import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.EntitySystem;
import com.badlogic.ashley.core.Family;
import com.badlogic.ashley.utils.ImmutableArray;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.utils.Array;
import com.questmark.entity.Mapper;
import com.questmark.entity.QuadTree;
import com.questmark.entity.components.*;

/**
 * A LibGDX Ashley {@link EntitySystem} that handles the collision between any two entities
 * via their {@link com.questmark.entity.components.BoundingBoxComponent} using axis aligned calculations.
 *
 * @author Ming Li
 */
public class EntityCollisionSystem extends EntitySystem implements CollisionSystem {

    private QuadTree quadTree;
    private Array<Rectangle> collisions;

    private ImmutableArray<Entity> entities;

    @Override
    public void addedToEngine(Engine engine) {
        entities = engine.getEntitiesFor(Family.all(BoundingBoxComponent.class, PositionComponent.class,
                VelocityComponent.class, PreviousPositionComponent.class).get());
        collisions = new Array<Rectangle>();
    }

    @Override
    public void update(float dt) {
        quadTree.clear();
        for (Entity e : entities) {
            BoundingBoxComponent b = Mapper.BOUNDING_BOX_MAPPER.get(e);
            quadTree.insert(b.bounds);
        }

        for (Entity e : entities) {
            BoundingBoxComponent bb = Mapper.BOUNDING_BOX_MAPPER.get(e);
            PositionComponent p = Mapper.POS_MAPPER.get(e);
            VelocityComponent v = Mapper.VEL_MAPPER.get(e);
            PreviousPositionComponent pp = Mapper.PREV_POS_MAPPER.get(e);

            collisions.clear();
            quadTree.retrieve(collisions, bb.bounds);

            for (Rectangle bounds : collisions) {
                if (!bb.bounds.equals(bounds)) {
                    if (bb.bounds.overlaps(bounds)) {
                        v.dx = v.dy = 0.f;
                        p.x = pp.x;
                        p.y = pp.y;
                    }
                }
            }
        }
    }

    @Override
    public void setMapData(int mapWidth, int mapHeight, int tileSize, Array<Rectangle> boundingBoxes) {
        quadTree = new QuadTree(0, new Rectangle(0, 0, mapWidth * tileSize, mapHeight * tileSize));
    }

}

File: core/src/com/questmark/entity/systems/collision/EntityCollisionSystem.kt
package com.questmark.entity.systems.collision

import com.badlogic.ashley.core.Engine
import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.core.EntitySystem
import com.badlogic.ashley.core.Family
import com.badlogic.ashley.utils.ImmutableArray
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.utils.Array
import com.questmark.entity.Mapper
import com.questmark.entity.QuadTree
import com.questmark.entity.components.*
import com.questmark.entity.components.enemy.EnemyComponent

/**
 * A LibGDX Ashley [EntitySystem] that handles the collision between any two entities
 * via their [com.questmark.entity.components.BoundingBoxComponent] using axis aligned calculations.
 *
 * @author Ming Li
 */
class EntityCollisionSystem : EntitySystem(), CollisionSystem {

    private var quadTree: QuadTree? = null
    private var collisions: Array<Rectangle>? = null

    private var entities: ImmutableArray<Entity>? = null

    override fun addedToEngine(engine: Engine?) {
        entities = engine!!.getEntitiesFor(Family.all(BoundingBoxComponent::class.java, PositionComponent::class.java,
                VelocityComponent::class.java, PreviousPositionComponent::class.java).exclude(EnemyComponent::class.java).get())
        collisions = Array()
    }

    override fun update(dt: Float) {
        quadTree!!.clear()
        for (e in entities!!) {
            val b = Mapper.BOUNDING_BOX_MAPPER!!.get(e)
            quadTree!!.insert(b.bounds)
        }

        for (e in entities!!) {
            val bb = Mapper.BOUNDING_BOX_MAPPER!!.get(e)
            val p = Mapper.POS_MAPPER!!.get(e)
            val v = Mapper.VEL_MAPPER!!.get(e)
            val pp = Mapper.PREV_POS_MAPPER!!.get(e)

            collisions!!.clear()
            quadTree!!.retrieve(collisions, bb.bounds)

            for (bounds in collisions!!) {
                if (bb.bounds != bounds) {
                    if (bb.bounds.overlaps(bounds)) {
                        v.v.set(0f, 0f)
                        p.p.set(pp.p.x, pp.p.y)
                    }
                }
            }
        }
    }

    override fun setMapData(mapWidth: Int, mapHeight: Int, tileSize: Int, boundingBoxes: Array<Rectangle>) {
        quadTree = QuadTree(0, Rectangle(0f, 0f, (mapWidth * tileSize).toFloat(), (mapHeight * tileSize).toFloat()))
    }

}
